Bridges mini rework - roadblocks

Sephirex95 opened the pull request:
![image](https://github.com/user-attachments/assets/ec1344fc-a35e-41ae-81aa-dd0aced61bbd)

Yes Patrick, bridges now act as a large roadblock stretching over the water, eliminating the need to place roadblocks just before or just after the bridge, and losing precious roaming walkers range!
![city 2025-06-08 23 02 08](https://github.com/user-attachments/assets/d7e80c15-2f19-4e98-bfec-f65a025f8bcd)
![city 2025-06-08 23 02 17](https://github.com/user-attachments/assets/ae7ff431-93bf-42a9-9d42-47b6beff4f0c)
![city 2025-06-08 23 02 21](https://github.com/user-attachments/assets/8260a6c4-a50d-4ef0-acd0-bf13775fef16)

Sephirex95 commented:
+260
-40

Yeah, like 10 lines per file, that's tame!

Keriew commented on src/window/building_info.c:
```diff
@@ -925,7 +936,8 @@ static int handle_specific_building_info_mouse(const mouse *m)
     } else if (context.figure.drawn) {
         return window_building_handle_mouse_figure_list(m, &context);
-    } else if (context.type == BUILDING_INFO_BUILDING) {
+
```
Unnecessary newline

Keriew commented on src/window/building_info.c:
```diff
@@ -286,7 +291,7 @@ static void init(int grid_offset)
     context.type = BUILDING_INFO_TERRAIN;
     context.figure.drawn = 0;
-    if (!context.building_id && map_sprite_bridge_at(grid_offset) > 0) {
+    if (map_is_bridge(grid_offset)) { //change this to is_bridge later after testing
```
I assume this comment is no longer needed.

Keriew reviewed (changes requested):
Some minor formatting changed requested to match the style.

Sephirex95 commented on src/window/building_info.c:
```diff
@@ -30,6 +30,7 @@
 #include "input/input.h"
 #include "map/aqueduct.h"
 #include "map/building.h"
+#include "map/bridge.h"
```
I had to recite alphabet twice out loud to understand.
r s t U
ah, right 
bridge before building...

Sephirex95 commented:
Fixed bugs and cleaned up formatting. 
To Do: 
Add missing tooltips to roadblock buttons in the bridge, fix the bug with flotsam blocking player from removing the bridge.

crudelios commented on src/map/bridge.c:
```diff
@@ -25,6 +31,10 @@ int map_bridge_building_length(void)
     return bridge.length;
 }
 
+int building_type_is_bridge(building_type type){
```
```suggestion
int building_type_is_bridge(building_type type)
{
```

crudelios commented on src/map/bridge.c:
```diff
@@ -223,25 +233,33 @@ int map_bridge_add(int x, int y, int is_ship_bridge)
+    int bridge_type = !is_ship_bridge ? BUILDING_LOW_BRIDGE : BUILDING_SHIP_BRIDGE;
+
+    building *b = building_create(bridge_type, x, y); 
+    //int prev_part = 0;
```
Remove please

crudelios commented on src/map/bridge.c:
```diff
@@ -223,25 +233,33 @@ int map_bridge_add(int x, int y, int is_ship_bridge)
     }
 
     int grid_offset = map_grid_offset(x, y);
+    int bridge_type = !is_ship_bridge ? BUILDING_LOW_BRIDGE : BUILDING_SHIP_BRIDGE;
```
```suggestion
    int bridge_type = is_ship_bridge ? BUILDING_SHIP_BRIDGE : BUILDING_LOW_BRIDGE;
```

Easier to follow

Sephirex95 replied:
there was a comment there 'reverse before PR', and I deleted the comment without reversing the logic.
Duh! :D

crudelios commented on src/map/bridge.h:
```diff
@@ -17,6 +20,10 @@ int map_bridge_get_sprite_id(int index, int length, int direction, int is_ship_b
  */
 int map_bridge_add(int x, int y, int is_ship_bridge);
 
+int get_y_bridge_tiles(int grid_offset);
+
+int get_x_bridge_tiles(int grid_offset);
```
```suggestion
int map_bridge_get_y_tiles(int grid_offset);

int map_bridge_get_x_tiles(int grid_offset);
```

crudelios commented on src/map/terrain.c:
```diff
@@ -469,6 +473,44 @@ static void determine_original_trees(buffer *images, int legacy_buffer)
     }
 }
 
+static int legacy_map_is_bridge(int grid_offset){ 
```
```suggestion
static int legacy_map_is_bridge(int grid_offset)
{ 
```

crudelios commented on src/map/terrain.c:
```diff
@@ -469,6 +473,44 @@ static void determine_original_trees(buffer *images, int legacy_buffer)
+    return (map_sprite_bridge_at(grid_offset)) && map_terrain_is(grid_offset, TERRAIN_WATER);
+}
+
+void map_terrain_migrate_old_bridges(void){
```
```suggestion
void map_terrain_migrate_old_bridges(void)
{
```

crudelios commented on src/map/terrain.c:
```diff
@@ -469,6 +473,44 @@ static void determine_original_trees(buffer *images, int legacy_buffer)
+    for (int y = 0; y < GRID_SIZE; y++) {
+        for (int x = 0; x < GRID_SIZE; x++) {
+            int grid_offset = map_grid_offset(x,y);
+            if (legacy_map_is_bridge(grid_offset) && !map_is_bridge(grid_offset)){ //old bridge but not new bridge
```
```suggestion
            if (legacy_map_is_bridge(grid_offset) && !map_is_bridge(grid_offset)) { //old bridge but not new bridge
```

crudelios commented on src/widget/city_with_overlay.c:
```diff
@@ -222,7 +222,10 @@ static void draw_flattened_building_footprint(const building *b, int x, int y, i
         image_base += 4;
     }
-    if (b->size == 1) {
+    if (building_type_is_bridge(b->type)){//dont draw bridges
+        return;
+    }
+    if (b->size == 1)  {
```
```suggestion
    if (building_type_is_bridge(b->type)) { // Don't draw bridges
        return;
    }
    if (b->size == 1) {
```

Extreme nitpicking

crudelios commented on src/window/building/figures.c:
```diff
@@ -30,6 +30,7 @@
 #include "translation/translation.h"
 #include "widget/city.h"
 #include "window/city.h"
+#include "utility.h"
```
```suggestion
#include "building/utility.h"
```

For clarity reasons, we always include the folder for headers that are not this files direct header.

Also, since we order headers alphabetically, this `#include` needs to move up.

crudelios commented on src/window/building/figures.c:
```diff
@@ -711,7 +712,15 @@ int window_building_handle_mouse_figure_list(const mouse *m, building_info_conte
             window_request_refresh();
         }
     }
+    if (c->terrain_type == TERRAIN_INFO_BRIDGE){
```
```suggestion
    if (c->terrain_type == TERRAIN_INFO_BRIDGE) {
```

I see a spacing pattern :wink:

crudelios commented on src/window/building/terrain.c:
```diff
@@ -69,5 +69,14 @@ void window_building_draw_terrain(building_info_context *c)
+        if (c->terrain_type == TERRAIN_INFO_BRIDGE){ //bridge
+            if(c->show_special_orders) {
+                window_building_draw_roadblock_orders(c);      
+            } else
+                window_building_draw_roadblock_button(c);
+        }
```
```suggestion
            } else {
                window_building_draw_roadblock_button(c);
            }
        }
```

crudelios commented on src/window/building/utility.c:
```diff
@@ -479,3 +518,19 @@ int window_building_handle_mouse_roadblock_orders(const mouse *m, building_info_
+
+int window_building_handle_mouse_bridge(const mouse *m, building_info_context *c){
+
+    if (window_building_handle_mouse_figure_list(m,c)){
```
```suggestion
    if (window_building_handle_mouse_figure_list(m,c)) {
```

crudelios commented on src/window/building/utility.c:
```diff
@@ -479,3 +518,19 @@ int window_building_handle_mouse_roadblock_orders(const mouse *m, building_info_
+        return 1;
+    }
+
+    if (c->show_special_orders) {
+            return window_building_handle_mouse_roadblock_orders(m, c);
+        } else {
+            return window_building_handle_mouse_roadblock_button(m, c);
+        }
```
Spacing issue

crudelios commented on src/window/building_info.c:
```diff
@@ -876,6 +881,12 @@ static void draw_foreground(void)
     } else if (context.type == BUILDING_INFO_LEGION) {
         window_building_draw_legion_info_foreground(&context);
+    } else if (context.terrain_type == TERRAIN_INFO_BRIDGE){
+            if (context.show_special_orders) {
+                window_building_draw_roadblock_orders_foreground(&context);
+            } else {
+                window_building_draw_roadblock_button(&context);
+            }
```
Tab spacing issue

crudelios reviewed (changes requested):
The bulk of the changes look fine, just some code style comments.

Sephirex95 commented:
Got mixed up with different branches. Pulling it and resubmitting
